### ISSUE TITLE
Envoy to forward application level spans to Tracing

### ISSUE THREAD

**Issue by Stono** [istio/istio#3395]
Hey, 
Considering istio is designed to have tight tracing integration with Jaeger, do we think we could add the udp -> jaeger collector forwarding capability of the jaeger-agent sidecar to envoy, to save us running an additional sidecar just for jaeger?

https://github.com/jaegertracing/jaeger-kubernetes#deploying-the-agent-as-sidecar

**Comment by sakshigoel12** [comment]
@douglas-reid can you help answer this question

**Comment by nmittler** [comment]
@mandarjog any thoughts?

**Comment by nmittler** [comment]
@douglas-reid @mandarjog gentle ping

**Comment by Stono** [comment]
The other driver for us here is to allow our applications to blindly fire udp packets to localhost rather than instantiating TCP connections to Jaeger for their spans.

**Comment by mandarjog** [comment]
@Stono I think it makes sense to add support directly to envoy. Running another sidecar will be non-starter. 
Does Jaeger support receiving UDP traces directly, or does it always have to be a bridge from localhost:udp --> remote jaeger:tcp ?

If it is the former, services can simply fire and forget traces to the remote UDP (like statsd)

If not, we will have to convince Envoy that adding an additional tracer / sink is beneficial for the wider community.

https://github.com/envoyproxy/envoy/tree/master/source/common/tracing

**Comment by douglas-reid** [comment]
Late to see this. I don't know that Istio has tight integration with Jaeger, but it is supported, mainly through the Zipkin-compatible endpoint.

As @mandarjog points out, the simplest way to use the UDP endpoint would be to add that functionality directly to Envoy.

**Comment by Stono** [comment]
Maybe tight was the wrong word, but I think you understand my point.  I believe you (istio) are trying to enable a pretty seamless integration with Jaeger and I'm just giving you a heads up that by using a model like the one in `jaeger-agent` as described in [their documentation](https://github.com/jaegertracing/jaeger-kubernetes#deploying-the-agent-as-sidecar) you enable `app -> localhost (udp) -> agent -> collector (tcp)`.

This means:

  - apps aren't creating tcp connections to jaeger
  - apps don't need configuration for the jaeger endpoint

Unfortunately jaeger doesn't support UDP traces directly (at least I don't believe it does), and to be honest even if it did - i'd still have to pass that level of configuration to the app to tell it where to send the UDP traces too.  I'm trying to think about this more from a platform perspective where apps don't need to know such information.

It probably makes sense to be in envoy as you say, I just don't have that relationship with them like I do with the istio team, and from my perspective i'm actually using `docker.io/istio/proxy`, not envoy, so that's why i raised the issue here.

**Comment by Stono** [comment]
Hey @douglas-reid this as probably another one we should have discussed today!  Chatting through it with Nathan we realised it's probably not worded the best...

**Comment by nmittler** [comment]
@douglas-reid I've temporarily assigned to you ... feel free to reassign as necessary.  It seems like this might be a generally useful feature.

**Comment by objectiser** [comment]
@Stono One solution to this would be to deploy the `jaeger-agent` as a daemonset, which is the approach we are now using in the [production deployments of Jaeger](https://github.com/jaegertracing/jaeger-kubernetes/blob/master/jaeger-production-template.yml#L167).

Someone is in the process of [updating the kubernetes jaeger helm chart](https://github.com/kubernetes/charts/issues/5677#issuecomment-392748735) to include the all-in-one configuration and deploy the agent in this way. Once that is done, we can use the same approach in the helm chart used to deploy Jaeger in istio.

This means you won't need any specific config for your app - it will report to the local agent which will then forward to the collector.

**Comment by Stono** [comment]
@objectiser I don't feel the need to deploy more applications in order to facilitate this though, it doesn't really make sense (unless i'm missing something).

Envoy is already sending traces/spans to the tracing service, it's already got all of that configuration from control plane and it's good to go.  I don't see why it couldn't also listen on localhost for spans being sent from the application too, and piggyback over the same mechanism.

This would actually then be a tracing provider agnostic way of sending the traces and spans.

**Comment by nmittler** [comment]
@douglas-reid @objectiser any updates?

**Comment by douglas-reid** [comment]
i don't have any updates here directly. Some potential relevant Envoy links: 
- https://github.com/envoyproxy/envoy/issues/492
- https://docs.google.com/document/d/1G9IVq7F7Onwinsl6EYzGsdzAGvVbo2FGfcPt35ItIx8/edit

cc: @cmluciano , as he has been doing some of the initial UDP in Envoy support design thinking.

On a different note, there is Mixer Direct API work kicking-off, in collaboration with the OpenCensus team.  Supporting the Jaeger API could be a priority, if that was an attractive option. But I'm not sure that would solve all of your ask (clients would still have to be configured).

**Comment by objectiser** [comment]
If the `jaeger-agent` was integrated into the envoy project, this would enable the service to seamlessly report spans via UDP (with no config in the service), but the spans would be using the native Jaeger data format.

Therefore the embedded `jaeger-agent` would need its own HTTP connection to the `jaeger-collector` to report the spans - thus creating two connections from each proxy for reporting tracing data.

The only way to reuse the existing connection from the proxy would be for the data to be reported in the zipkin format, which would require a suitable port to be exposed on the proxy and the service to be configured to use the `ZipkinSender`. So does not meet the requirement for zero-config in the service.

Once https://github.com/envoyproxy/envoy/issues/492 is implemented, it may be possible to have a UDP to TCP proxy that can report Jaeger span data indirectly to the `jaeger-collector`, but this would still create the second connection for reporting tracing data.

Only other option may be to use the Jaeger native tracer in envoy, and then piggy back the spans from the UDP port on that connection somehow.

**Comment by Stono** [comment]
Not sure about this bot, this is a valid feature request IMO but I appreciate it that it's low priority?

**Comment by Stono** [comment]
Same with this one @howardjohn @geeknoid

**Comment by Stono** [comment]
@howardjohn :'(

**Comment by howardjohn** [comment]
:confused:
